frontend/src/components/Agreements/AgreementDetails/AgreementHistoryPanel.jsx

```diff
@@ -0,0 +1,173 @@
+import PropTypes from "prop-types";
+import { useGetAgreementHistoryByIdQuery } from "../../../api/opsAPI";
+import { historyData } from "../../../pages/agreements/details/data";
+import LogItem from "../../UI/LogItem";
+import { convertCodeForDisplay } from "../../../helpers/utils";
+
+const findObjectTitle = (historyItem) => {
+    if (historyItem.class_name === "BudgetLineItem") {
+        return historyItem.event_details.line_description;
+    } else {
+        return historyItem.event_details.name;
+    }
+};
+
+const summaryMessage = (historyItem) => {
+    const className = convertCodeForDisplay("className", historyItem.class_name);
+    const user_name = historyItem.created_by_user_full_name;
+    const objectTitle = findObjectTitle(historyItem);
+    const classAndTitle = `${className}, "${objectTitle}",`;
+    if (historyItem.event_type === "NEW") {
+        return `New ${classAndTitle} created by ${user_name}.`;
+    } else if (historyItem.event_type === "UPDATED") {
+        return `${classAndTitle} updated by ${user_name}.`;
+    } else if (historyItem.event_type === "DELETED") {
+        return `${classAndTitle} deleted by ${user_name}.`;
+    }
+    return `${className} ${historyItem.event_type} ${user_name}`;
+};
+
+const getPropertyLabel = (className, fieldName) => {
+    if (className === "BudgetLineItem") return convertCodeForDisplay("budgetLineItemPropertyLabels", fieldName);
+    return convertCodeForDisplay("agreementPropertyLabels", fieldName);
+};
+
+const usersToNames = (users) => {
+    return users.map((user) => user.full_name);
+};
+
+const ChangesDetails = ({ historyItem }) => {
+    const changes = historyItem.changes;
+    const eventType = historyItem.event_type;
+    if (eventType != "UPDATED") return;
+    console.log("ChangesDetails.changes:", changes);
+    const prepChanges = Object.entries(changes).map(([key, change]) => {
+        if ("collection_of" in change) {
+            const added = change.collection_of == "User" ? usersToNames(change.added) : change.added;
+            const deleted = change.collection_of == "User" ? usersToNames(change.deleted) : change.deleted;
+            return {
+                propertyLabel: getPropertyLabel(historyItem.class_name, key),
+                isCollection: true,
+                added: added,
+                deleted: deleted,
+            };
+        } else if (key === "procurement_shop_id") {
+            const new_val = historyItem.event_details?.procurement_shop?.name;
```

Comment:
Suggested Refactor for future updates: 
I could see breaking this bit out into a function like `generateChangeObject` or something to `DRY` it out a bit. This could also help reduce the number of `if/else if`, since there's a clear correlation between the `key === {name}` and the `event_details.{property}`, so a simple mapping lookup or something, to do actual output, since the logic is the same across most.

Reply:
yeah, it can be improved.  I got part way into an idea and reverted to `if/else-if` as I wondered if maybe a little backend work could help simplify this.

Two things keep this from being almost a universal one liner, not just here but anywhere we want to turn a FK ID into a display string.  All but `project_officer`, use `<related_object_name>_id` for the FK ID.  And then the property name to display varies for classes -  User -> `full_name`, Agreement -> `name`, ResearchProject -> `title`, BudgetLineItem -> `line_description`.  So I am wondering if maybe there should be some back end changes first.  Changing project_officer to project_officer_id would be a breaking change and the project_officer user isn't included anyway - so maybe not that.  But can we add something to the base model, say `display_name` with a getter implemented on each class in a way that it's included in the JSON without messing with the ORM/DB?  If so, I could also use this for the summary message to say which Agreement/BLI was edited and use it for common `to_slim_dict()` - just ID + display_name.